[AAP] Add GKE docs (#36156)
DataDog/documentation: content/en/security/application_security/setup/kubernetes/gke.md

code_lang: gke
code_lang_weight: 50
further_reading:
  - link: "https://www.datadoghq.com/blog/app-api-protection-envoy-istio-nginx-haproxy"
    tag: "Blog"
    text: "Secure your APIs through Envoy, Istio, NGINX, HAProxy, and more with Datadog App and API Protection"
  - link: 'https://docs.cloud.google.com/kubernetes-engine/docs/how-to/configure-gke-service-extensions'
    tag: "Documentation"
    text: "Configure Service Extensions on a GKE Gateway"
  - link: 'https://cloud.google.com/service-extensions/docs/overview'
    tag: "Documentation"
    text: "Google Cloud Service Extensions overview"
  - link: "/security/default_rules/?category=cat-application-security"
    tag: "Documentation"
    text: "OOTB App and API Protection Rules"
  - link: "/security/application_security/troubleshooting"
    tag: "Documentation"
    text: "Troubleshooting App and API Protection"
  - link: 'https://github.com/DataDog/dd-trace-go/tree/main/contrib/envoyproxy/go-control-plane/cmd/serviceextensions'
    tag: "Source Code"
    text: "App and API Protection Service Extension source code"
---

{{< callout url="#" btn_hidden="true" header="App and API Protection for GKE is in Preview" >}}

> To try the preview of App and API Protection for GKE, use the following setup instructions.

Reviewer: ```suggestion
To try the Preview of App and API Protection for GKE, use the following setup instructions.
```

{{< /callout >}}

You can enable Datadog [App and API Protection][1] on a Google Kubernetes Engine (GKE) Gateway to inspect and protect traffic at the cluster edge. The integration uses the GKE Gateway controller and the `GCPTrafficExtension` custom resource to attach the Datadog security processor to the GCP Cloud Load Balancer.

## Prerequisites

- A GKE cluster running **version 1.33 or later** with the [Gateway API enabled][2].
- A `Gateway` resource using a [GatewayClass][3] that supports traffic extensions, for example `gke-l7-global-external-managed` or `gke-l7-regional-external-managed`. For the full list, see [supported GatewayClasses for `GCPTrafficExtension`][12].
- The [Datadog Agent is installed and configured][4] in your Kubernetes cluster.
  - Enable and configure [Remote Configuration][5] to block attackers through the Datadog UI.

>   - Enable [APM][6] in the Agent so the security processor can send its own traces to the Agent.

Reviewer: ```suggestion
- A GKE cluster running **version 1.33 or later**, with the [Gateway API enabled][2].
- A `Gateway` resource that uses a [GatewayClass][3] supporting traffic extensions, such as `gke-l7-global-external-managed` or `gke-l7-regional-external-managed`. For the full list, see [supported `GatewayClasses` for `GCPTrafficExtension`][12].
- The [Datadog Agent is installed and configured][4] in your Kubernetes cluster.
  - [Remote Configuration][5] is enabled and configured in the Datadog UI to block attackers.
  - [APM][6] is enabled in the Agent so the security processor can send its own traces to the Agent.
```

## Enabling threat detection

### Overview

> 

Reviewer: ```suggestion
```

            - name: DD_AGENT_HOST
              value: "<your-datadog-agent-service>.<your-datadog-agent-namespace>.svc.cluster.local"
            - name: DD_TRACE_AGENT_PORT # Optional if your Agent's trace port is the default 8126
              value: "8126"
          readinessProbe:
            httpGet:
              path: /
              port: health
            initialDelaySeconds: 5
            periodSeconds: 10
          livenessProbe:
            httpGet:
              path: /
              port: health
            initialDelaySeconds: 15
            periodSeconds: 20
---
apiVersion: v1
kind: Service
metadata:
  name: datadog-aap-extproc-service # This name is referenced by the GCPTrafficExtension
  namespace: <your-gateway-namespace> # Same namespace as the Gateway
  labels:
    app: datadog-aap-extproc
spec:
  selector:
    app: datadog-aap-extproc
  ports:
    - name: grpc
      port: 443
      targetPort: grpc
      protocol: TCP
      appProtocol: HTTP2 # Required: GKE callout backends must use HTTP/2
    - name: health
      port: 80
      targetPort: health
      protocol: TCP
```

<div class="alert alert-info">

>   <strong>Note:</strong> On GKE, backends used as Service Extension callouts must speak HTTP/2. The <code>appProtocol: HTTP2</code> field on the <code>grpc</code> Service port tells the GKE Gateway controller to use HTTP/2 when forwarding traffic. Without it, the load balancer cannot talk gRPC to the security processor.

Reviewer: ```suggestion
  On GKE, backends used for Service Extension callouts must support HTTP/2. Setting <code>appProtocol: HTTP2</code> on the <code>gRPC</code> Service port tells the GKE Gateway controller to forward traffic over HTTP/2. Without it, the load balancer can't send gRPC traffic to the security processor.
```

          authority: "datadog-aap-extproc-service.<your-gateway-namespace>.svc.cluster.local"

          # By default, if the security processor fails or times out, the proxy returns a
          # 5xx error. To prevent this, enable the failOpen setting. When enabled, request
          # processing is stopped on error but the request is not dropped, which keeps the
          # application available.
          failOpen: true

          # Mandatory: Only set the Request and Response Headers events.
          # If body events are selected, the security processor transfers and analyzes
          # the bodies for every request without applying the relevant selection rules.
          # Use the DD_APPSEC_BODY_PARSING_SIZE_LIMIT environment variable on the callout
          # container to enable body processing.
          supportedEvents:
            - RequestHeaders
            - ResponseHeaders

          # Adjust the timeout to match your processing needs (for example, with body processing).
          # Valid range: 10ms to 10s. This is the same option as message_timeout in an Envoy ext_proc configuration.
          timeout: 1s
```

<div class="alert alert-info">

>   <strong>Note:</strong> The <code>Gateway</code>, the <code>GCPTrafficExtension</code>, and the security processor <code>Service</code> referenced by <code>backendRef</code> must all reside in the same namespace. See the <a href="https://cloud.google.com/kubernetes-engine/docs/how-to/configure-gke-service-extensions">GKE Service Extensions restrictions</a> for the full list of constraints on the extension's backend <code>Service</code>.

Reviewer: ```suggestion
  The <code>Gateway</code>, the <code>GCPTrafficExtension</code>, and the security processor <code>Service</code> referenced by <code>backendRef</code> must all reside in the same namespace. See the <a href="https://cloud.google.com/kubernetes-engine/docs/how-to/configure-gke-service-extensions">GKE Service Extensions restrictions</a> for the full list of constraints on the extension's backend <code>Service</code>.
```
